library/alloc/src/vec/spec_extend.rs

```diff
@@ -57,31 +100,23 @@ where
     }
 }
 
-impl<T, A: Allocator> SpecExtend<T, IntoIter<T>> for Vec<T, A> {
-    fn spec_extend(&mut self, mut iterator: IntoIter<T>) {
-        unsafe {
-            self.append_elements(iterator.as_slice() as _);
-        }
-        iterator.ptr = iterator.end;
-    }
-}
-
-impl<'a, T: 'a, I, A: Allocator + 'a> SpecExtend<&'a T, I> for Vec<T, A>
+impl<T, I, A: Allocator> SpecExtendInner<T, I> for Vec<T, A>
 where
-    I: Iterator<Item = &'a T>,
-    T: Clone,
+    I: TrustedLen<Item = T> + TrustedRandomAccess,
 {
-    default fn spec_extend(&mut self, iterator: I) {
-        self.spec_extend(iterator.cloned())
-    }
-}
+    default fn spec_extend(&mut self, mut iterator: I) {
+        let size = iterator.size();
+        self.reserve(size);
 
-impl<'a, T: 'a, A: Allocator + 'a> SpecExtend<&'a T, slice::Iter<'a, T>> for Vec<T, A>
-where
-    T: Copy,
-{
-    fn spec_extend(&mut self, iterator: slice::Iter<'a, T>) {
-        let slice = iterator.as_slice();
-        unsafe { self.append_elements(slice) };
+        // SAFETY: reserve ensured that there is sufficient capacity for the additional items.
+        // We use ptr.add() instead of offset() to avoid problems with ZSTs.
```

Comment:
What is the intention here? raw pointer .add() and .offset() handle ZSTs the same way

Reply:
A Vec of ZSTs can have a length greater than `isize::MAX`.

Reply:
I see, I think I know what you mean then. Whether calling add or offset, it shouldn't matter given that add is implemented in terms of offset. It should be correct with either of them (since the offset *in bytes* will be zero anyway).